### PR TITLE
Microsoft windows defender ATP Add new action Get Missing Software Updates

### PR DESCRIPTION
**Pull request by r7-mgorny**
## Proposed Changes
Microsoft windows defender ATP Add new action Get Missing Software Updates

### Description

Describe the proposed changes:

  -

### Testing

Any testing information goes here.

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Style

Review the [style guide](https://komand.github.io/python/style.html)

- [ ] For dependencies, pin [OS package](https://komand.github.io/python/style.html#dockerfile) and [Python package](https://komand.github.io/python/style.html#requirements-txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://komand.github.io/python/sdk.html#sdk-versions) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://komand.github.io/python/error_handling.html#plugin-exceptions) and [ConnectionTestException](https://komand.github.io/python/error_handling.html#connection-exceptions)
- [ ] For logging, use [self.logger](https://komand.github.io/python/sdk.html#logging)
- [ ] For docs, use [changelog style](https://komand.github.io/python/style.html#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://komand.github.io/python/style.html#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{
  "body": {
    "log": "Connecting...\nTime Now: 1600041857.711843\nTime Ago: 0\nSeconds elapsed:1600041857\nRefreshing auth token\nUpdating Auth Token...\nGetting token from: https://login.windows.net/5c824599-dc8c-4d31-96fb-3b886d4f8f10/oauth2/token\nAuthentication was successful, token is: ******************D5WGg\nrapid7/Microsoft Windows Defender ATP:4.4.1. Step name: get_missing_software_updates\nTime Now: 1600041858.0871747\nTime Ago: 1600041858.084996\nSeconds elapsed:0\nToken is valid, not refreshing.\nUpdating session headers.\nAttempting to isolate machine id: a542d20dceff077a189b84f8871747d674fb57a2\nTime Now: 1600041859.5862575\nTime Ago: 1600041858.084996\nSeconds elapsed:1\nToken is valid, not refreshing.\nUpdating session headers.\n",
    "meta": {},
    "output": {
      "updates": [
        {
          "cveAddressed": 68,
          "id": "4570333",
          "machineMissedOn": 1,
          "name": "September 2020 Security Updates",
          "osBuild": 17763,
          "productsNames": [
            "windows_10",
            "internet_explorer",
            "edge"
          ],
          "url": "https://catalog.update.microsoft.com/v7/site/Search.aspx?q=KB4570333"
        }
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/microsoft_atp:4.4.1 --debug run < tests/get_missing_software_updates_by_hostname.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connecting...\nTime Now: 1600041865.14162\nTime Ago: 0\nSeconds elapsed:1600041865\nRefreshing auth token\nUpdating Auth Token...\nGetting token from: https://login.windows.net/5c824599-dc8c-4d31-96fb-3b886d4f8f10/oauth2/token\nAuthentication was successful, token is: ******************K7X1Q\nrapid7/Microsoft Windows Defender ATP:4.4.1. Step name: get_missing_software_updates\nTime Now: 1600041865.4857838\nTime Ago: 1600041865.483377\nSeconds elapsed:0\nToken is valid, not refreshing.\nUpdating session headers.\nAttempting to isolate machine id: a542d20dceff077a189b84f8871747d674fb57a2\nTime Now: 1600041865.8911161\nTime Ago: 1600041865.483377\nSeconds elapsed:0\nToken is valid, not refreshing.\nUpdating session headers.\n",
    "meta": {},
    "output": {
      "updates": [
        {
          "cveAddressed": 68,
          "id": "4570333",
          "machineMissedOn": 1,
          "name": "September 2020 Security Updates",
          "osBuild": 17763,
          "productsNames": [
            "windows_10",
            "internet_explorer",
            "edge"
          ],
          "url": "https://catalog.update.microsoft.com/v7/site/Search.aspx?q=KB4570333"
        }
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/microsoft_atp:4.4.1 --debug run < tests/get_missing_software_updates_by_id.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connecting...\nTime Now: 1600041870.2531648\nTime Ago: 0\nSeconds elapsed:1600041870\nRefreshing auth token\nUpdating Auth Token...\nGetting token from: https://login.windows.net/5c824599-dc8c-4d31-96fb-3b886d4f8f10/oauth2/token\nAuthentication was successful, token is: ******************fHT_w\nrapid7/Microsoft Windows Defender ATP:4.4.1. Step name: get_missing_software_updates\nTime Now: 1600041870.5777783\nTime Ago: 1600041870.5764802\nSeconds elapsed:0\nToken is valid, not refreshing.\nUpdating session headers.\nAttempting to isolate machine id: a542d20dceff077a189b84f8871747d674fb57a2\nTime Now: 1600041871.6314468\nTime Ago: 1600041870.5764802\nSeconds elapsed:1\nToken is valid, not refreshing.\nUpdating session headers.\n",
    "meta": {},
    "output": {
      "updates": [
        {
          "cveAddressed": 68,
          "id": "4570333",
          "machineMissedOn": 1,
          "name": "September 2020 Security Updates",
          "osBuild": 17763,
          "productsNames": [
            "windows_10",
            "internet_explorer",
            "edge"
          ],
          "url": "https://catalog.update.microsoft.com/v7/site/Search.aspx?q=KB4570333"
        }
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/microsoft_atp:4.4.1 --debug run < tests/get_missing_software_updates_by_ip.json
</summary>
</details>

### Validate

<details>

```
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../tools/Makefiles/Helpers.mk ../tools/Makefiles/Colors.mk
--
[*] Running validators
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Plugin successfully validated!

----
[*] Total time elapsed: 2717.526ms

[*] Validating spec with js-yaml
[SUCCESS] Passes js-yaml spec check


[*] For Markdown linting, install mdl: gem install mdl

[*] Validating python files for style...
./komand_microsoft_atp/triggers/get_alerts/trigger.py:6:1: E302 expected 2 blank lines, found 1
./komand_microsoft_atp/triggers/get_alerts/trigger.py:29:68: E261 at least two spaces before inline comment
./komand_microsoft_atp/triggers/get_alerts/trigger.py:46:34: F541 f-string is missing placeholders
./komand_microsoft_atp/triggers/get_alert_matching_key/trigger.py:31:68: E261 at least two spaces before inline comment
./komand_microsoft_atp/triggers/get_alert_matching_key/trigger.py:59:34: F541 f-string is missing placeholders
./komand_microsoft_atp/util/api.py:85:42: F541 f-string is missing placeholders
./komand_microsoft_atp/util/api.py:196:9: F841 local variable 'e' is assigned to but never used
./komand_microsoft_atp/util/api.py:198:9: F841 local variable 'e' is assigned to but never used
./unit_test/test_get_alerts.py:5:1: E402 module level import not at top of file
./unit_test/test_get_alerts.py:6:1: E402 module level import not at top of file
./unit_test/test_get_alerts.py:7:1: E402 module level import not at top of file
./unit_test/test_get_alerts.py:8:1: E402 module level import not at top of file
./unit_test/test_get_alerts.py:9:1: E402 module level import not at top of file
./unit_test/test_get_alerts.py:10:1: E402 module level import not at top of file
./unit_test/test_get_alerts.py:11:1: E402 module level import not at top of file
./unit_test/test_get_alerts.py:28:1: E302 expected 2 blank lines, found 1
./unit_test/test_get_alerts.py:33:1: E302 expected 2 blank lines, found 1
./unit_test/test_get_alerts.py:46:9: F841 local variable 'e' is assigned to but never used
./unit_test/test_get_alerts.py:49:1: W293 blank line contains whitespace
./unit_test/test_get_alerts.py:50:110: W291 trailing whitespace
./unit_test/test_get_alerts.py:65:20: E261 at least two spaces before inline comment
./unit_test/test_get_alerts.py:66:1: W391 blank line at end of file
./unit_test/test_get_alert_matching_key.py:5:1: E402 module level import not at top of file
./unit_test/test_get_alert_matching_key.py:6:1: E402 module level import not at top of file
./unit_test/test_get_alert_matching_key.py:7:1: E402 module level import not at top of file
./unit_test/test_get_alert_matching_key.py:8:1: E402 module level import not at top of file
./unit_test/test_get_alert_matching_key.py:9:1: E402 module level import not at top of file
./unit_test/test_get_alert_matching_key.py:10:1: E402 module level import not at top of file
./unit_test/test_get_alert_matching_key.py:11:1: E402 module level import not at top of file
./unit_test/test_get_alert_matching_key.py:28:1: E302 expected 2 blank lines, found 1
./unit_test/test_get_alert_matching_key.py:33:1: E302 expected 2 blank lines, found 1
./unit_test/test_get_alert_matching_key.py:55:9: F841 local variable 'e' is assigned to but never used
./unit_test/test_get_alert_matching_key.py:58:1: W293 blank line contains whitespace
./unit_test/test_get_alert_matching_key.py:59:110: W291 trailing whitespace
./unit_test/test_get_alert_matching_key.py:74:20: E261 at least two spaces before inline comment
./unit_test/test_get_alert_matching_key.py:82:40: W292 no newline at end of file
./unit_test/test_get_file_id_from_alert_id.py:5:1: E402 module level import not at top of file
./unit_test/test_get_file_id_from_alert_id.py:6:1: E402 module level import not at top of file
./unit_test/test_get_file_id_from_alert_id.py:7:1: E402 module level import not at top of file
./unit_test/test_get_file_id_from_alert_id.py:8:1: E402 module level import not at top of file
./unit_test/test_get_file_id_from_alert_id.py:9:1: E402 module level import not at top of file
./unit_test/test_get_file_id_from_alert_id.py:26:9: F841 local variable 'e' is assigned to but never used
./unit_test/test_get_file_id_from_alert_id.py:29:1: W293 blank line contains whitespace
./unit_test/test_get_file_id_from_alert_id.py:30:110: W291 trailing whitespace
./unit_test/test_get_file_id_from_alert_id.py:36:9: E303 too many blank lines (2)
./unit_test/test_get_machine_action.py:6:1: E402 module level import not at top of file
./unit_test/test_get_machine_action.py:7:1: E402 module level import not at top of file
./unit_test/test_get_machine_action.py:8:1: E402 module level import not at top of file
./unit_test/test_get_machine_action.py:9:1: E402 module level import not at top of file
./unit_test/test_get_machine_action.py:10:1: E402 module level import not at top of file
./unit_test/test_get_machine_action.py:40:9: F841 local variable 'e' is assigned to but never used
./unit_test/test_get_machine_action.py:44:110: W291 trailing whitespace
./unit_test/test_get_machine_action.py:57:69: W291 trailing whitespace
./unit_test/test_get_machine_action.py:72:45: W292 no newline at end of file
./unit_test/test_get_machine_id_from_alert_id.py:5:1: E402 module level import not at top of file
./unit_test/test_get_machine_id_from_alert_id.py:6:1: E402 module level import not at top of file
./unit_test/test_get_machine_id_from_alert_id.py:7:1: E402 module level import not at top of file
./unit_test/test_get_machine_id_from_alert_id.py:8:1: E402 module level import not at top of file
./unit_test/test_get_machine_id_from_alert_id.py:9:1: E402 module level import not at top of file
./unit_test/test_get_machine_id_from_alert_id.py:26:9: F841 local variable 'e' is assigned to but never used
./unit_test/test_get_machine_id_from_alert_id.py:29:1: W293 blank line contains whitespace
./unit_test/test_get_machine_id_from_alert_id.py:30:110: W291 trailing whitespace
./unit_test/test_get_machine_id_from_alert_id.py:36:9: E303 too many blank lines (2)
./unit_test/test_get_machine_information.py:5:1: E402 module level import not at top of file
./unit_test/test_get_machine_information.py:6:1: E402 module level import not at top of file
./unit_test/test_get_machine_information.py:7:1: E402 module level import not at top of file
./unit_test/test_get_machine_information.py:8:1: E402 module level import not at top of file
./unit_test/test_get_machine_information.py:9:1: E402 module level import not at top of file
./unit_test/test_get_machine_information.py:26:9: F841 local variable 'e' is assigned to but never used
./unit_test/test_get_machine_information.py:29:1: W293 blank line contains whitespace
./unit_test/test_get_machine_information.py:30:110: W291 trailing whitespace
./unit_test/test_get_machine_information.py:36:9: E303 too many blank lines (2)
./unit_test/test_isolate_machine.py:5:1: E402 module level import not at top of file
./unit_test/test_isolate_machine.py:6:1: E402 module level import not at top of file
./unit_test/test_isolate_machine.py:7:1: E402 module level import not at top of file
./unit_test/test_isolate_machine.py:8:1: E402 module level import not at top of file
./unit_test/test_isolate_machine.py:9:1: E402 module level import not at top of file
./unit_test/test_isolate_machine.py:26:9: F841 local variable 'e' is assigned to but never used
./unit_test/test_isolate_machine.py:29:1: W293 blank line contains whitespace
./unit_test/test_isolate_machine.py:30:110: W291 trailing whitespace
./unit_test/test_isolate_machine.py:36:9: E303 too many blank lines (2)
./unit_test/test_run_antivirus_scan.py:5:1: E402 module level import not at top of file
./unit_test/test_run_antivirus_scan.py:6:1: E402 module level import not at top of file
./unit_test/test_run_antivirus_scan.py:7:1: E402 module level import not at top of file
./unit_test/test_run_antivirus_scan.py:8:1: E402 module level import not at top of file
./unit_test/test_run_antivirus_scan.py:9:1: E402 module level import not at top of file
./unit_test/test_run_antivirus_scan.py:39:9: F841 local variable 'e' is assigned to but never used
./unit_test/test_run_antivirus_scan.py:42:1: W293 blank line contains whitespace
./unit_test/test_run_antivirus_scan.py:43:110: W291 trailing whitespace
./unit_test/test_run_antivirus_scan.py:49:9: E303 too many blank lines (2)
./unit_test/test_run_antivirus_scan.py:57:69: W291 trailing whitespace
./unit_test/test_run_antivirus_scan.py:72:45: W292 no newline at end of file
./unit_test/test_stop_and_quarantine_file.py:5:1: E402 module level import not at top of file
./unit_test/test_stop_and_quarantine_file.py:6:1: E402 module level import not at top of file
./unit_test/test_stop_and_quarantine_file.py:7:1: E402 module level import not at top of file
./unit_test/test_stop_and_quarantine_file.py:8:1: E402 module level import not at top of file
./unit_test/test_stop_and_quarantine_file.py:9:1: E402 module level import not at top of file
./unit_test/test_stop_and_quarantine_file.py:39:9: F841 local variable 'e' is assigned to but never used
./unit_test/test_stop_and_quarantine_file.py:42:1: W293 blank line contains whitespace
./unit_test/test_stop_and_quarantine_file.py:43:110: W291 trailing whitespace
./unit_test/test_stop_and_quarantine_file.py:49:9: E303 too many blank lines (2)
./unit_test/test_stop_and_quarantine_file.py:57:69: W291 trailing whitespace
./unit_test/test_stop_and_quarantine_file.py:72:45: W292 no newline at end of file
./unit_test/test_unisolate_machine.py:5:1: E402 module level import not at top of file
./unit_test/test_unisolate_machine.py:6:1: E402 module level import not at top of file
./unit_test/test_unisolate_machine.py:7:1: E402 module level import not at top of file
./unit_test/test_unisolate_machine.py:8:1: E402 module level import not at top of file
./unit_test/test_unisolate_machine.py:9:1: E402 module level import not at top of file
./unit_test/test_unisolate_machine.py:39:9: F841 local variable 'e' is assigned to but never used
./unit_test/test_unisolate_machine.py:42:1: W293 blank line contains whitespace
./unit_test/test_unisolate_machine.py:43:110: W291 trailing whitespace
./unit_test/test_unisolate_machine.py:49:9: E303 too many blank lines (2)
./unit_test/test_unisolate_machine.py:57:69: W291 trailing whitespace
./unit_test/test_unisolate_machine.py:72:45: W292 no newline at end of file
[FAIL] Fails flake8 linting

[*] Validating python files for security vulnerabilities...
[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: None
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
[main]  INFO    running on Python 3.8.5
61 [0.. 50.. ]
Run started:2020-09-14 00:09:30.536924

Test results:
        No issues identified.

Code scanned:
        Total lines of code: 3269
        Total lines skipped (#nosec): 0

Run metrics:
        Total issues (by severity):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
        Total issues (by confidence):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
Files skipped (0):
[SUCCESS] Passes bandit security checks

[*] Checking for typos with Misspell...
[SUCCESS] Passes Misspell check

```

<summary>
make validate
</summary>
</details>

<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
WARNING: URLs found that return a 4xx code. Verify they are publicly accessible and if not, update with a working URL.
violation: help.md[208]: https://api.securitycenter.windows.com/api/$metadata#Indicators/$entity
[*] Plugin successfully validated!

----
[*] Total time elapsed: 70222.921ms

```

<summary>
icon-validate --all .
</summary>
</details>

### CODE REVIEW

File: microsoft_atp/komand_microsoft_atp/actions/get_missing_software_updates/action.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    def __init__(self):
+        super(self.__class__, self).__init__(
+                name='get_missing_software_updates',
+                description=Component.DESCRIPTION,
+                input=GetMissingSoftwareUpdatesInput(),
+                output=GetMissingSoftwareUpdatesOutput())
+
+    def run(self, params={}):
+        machine_id = self.connection.client.find_first_machine(params.get(Input.MACHINE)).get("id")
+        self.logger.info("Attempting to isolate machine id: " + machine_id)
```
Comment (jschipp-r7) [inline microsoft_atp/komand_microsoft_atp/actions/get_missing_software_updates/action.py:17]: This log line doesn't make sense in this action. It sounds like we're trying to Isolate/Quarantine instead of getting software updates

```suggestion
        self.logger.info("Attempting to isolate machine id: " + machine_id)
```

**Review by jschipp-r7 — CHANGES_REQUESTED** [review]
